[WIP] Common error message helper.

kbrock opened the pull request:
## Overview.

This is the 3rd attempt.

Depends upon:

- [ ] #229 

Goals

- Provide more verbose error messages.
- Introduce more pedantic error checking.
- Move more error checking to initialize time.

## Changes

#### Parsing

- Ensure required fields are defined (see note).
- Reject fields that are the wrong type.
- Reject unknown fields.

#### Next processing

- Ensure `Next`/`Default` are valid.
- Add `Next` validation to `ChoiceRule` and `Catcher`.
- Reject a invalid `Next` field in a child `ChoiceRule`.
- Remove `State` access of `Workflow#payload["States"]` for validations (punted).

#### Reduction of scope

I did the following changes to reduce the scope of this

- No longer type checking/requiring on every field (booleans, timestamps, strings).
- No longer rejecting unknown fields.
- Fixes to Choice have been extracted to #189

I can easily add the first 2 items back but I was trying to keep this down.

## Before / After

```diff
- Missing field "States"
+ Workflow requires Hash field "States"
- Missing field "StartAt"
+ Workflow requires String field "StartAt"
- "StartAt" not in the "States" field
+ Workflow requires field "StartAt" to be in "States" list but got [Foo]
- Choice state must have "Choices"
+ States.Choice1 requires Array field "Choices"
- "Choices" must be a non-empty array
+ States.Choice1 requires Array field "Choices"
- "Default" not in "States"
+ States.Choice1 requires field "Default" to be in "States" list but got [MissingState]
- Invalid state type: [Invalid]
+ States.FirstState requires String field "Type" but got invalid value [Invalid]
- Missing "Type" field in state [FirstState]
+ States.FirstState requires String field "Type"
- (ruby runtime error)
+ States.FirstMatchState.Choices.0.Data requires Path field "Variable"
- Path [wrong] must start with "$"
+ States.FirstMatchState.Choices.0.Data requires Path field "Variable" Path [wrong] must start with "$"
- (ignored)
+ States.FirstMatchState.Choices.0.Not.0.Data does not recognize field "Next"
```

## Changelog History

1. Pass a Validator instead of `workflow`.
2. Pass an Enhance payload with getter methods.
3. `State#include Validation`. **<==**

Have another branch that:
- Adds booleans, timestamps, and strings back.
- Rejects unknown fields
- uses a dsl

kbrock commented:
update:
- introduced common validation helper methods

update:
- rubocops

kbrock commented:
update:
- trying to reduce complexity

kbrock commented:
update:
- extra use of `validate_field`
- remove validation methods with a single line.

kbrock commented:
update:
- Introduce a validator (so we won't be passing around the workflow everywhere)
- NonTerminal now runs own validation and sets next/end

kbrock commented:
update:
- implemented common error validation for other PRs
- separate validator from workflow (will pass around)

update:
- rubocop fixes

update:
- rubocop fixes

kbrock commented:
update:
- no longer able to pass state into validator
- validator `state_for` no longer modifies selt.

update:
- remove no longer valid validator comments
- string interpolation

kbrock commented:
update:
- rebase (to help other PRs resolve conflicts)

agrare commented on lib/floe/workflow/states/wait.rb:
```diff
@@ -8,22 +8,18 @@ module States
         def initialize(workflow, name, payload)
           super
 
-          @next           = payload["Next"]
-          @end            = !!payload["End"]
```
Why delete these two?  Per https://states-language.net/#state-type-table Wait requires either Next or End:true

agrare replied:
Oh I see you moved these into NonTerminalMixin

kbrock replied:
Good:
- everything dealing with Next/End is in one place
- No contract between the Mixin and the including class (currently, it assumes `@next` and `@end` are defined and set.
- Validation is no longer a contract between the two layers.

Bad:
- I really like the old way of having all variables explicitly stated in one place.

I guess I can change the way we parse it in `Wait#initialize` - that should fix it.

kbrock replied:
Got rid of this change

kbrock commented:
update:
- rebase
- no longer moving `End` and `Next` parsing in parent `NonTerminalMixin`
- validating Catch#`Next` is a valid ref, and `Catch#ErrorEquals` is a list
- validating `ChoiceRule` `Next` (first and second levels)
- validating `Retrier#ErrorEquals` is a list
- Fixed up Choice texts that were not valid

kbrock commented:
update:
- dropped `key?` to make code climate happy

kbrock commented:
Goal is to simplify this PR

update:
- rebase
- build upon name_path
- dropped PayloadValidator that wrapped a payload, and instead have a mixin.
- don't handle strings, booleans, timestamps, nor detect missing fields.

kbrock commented:
update:
- trying to get the complexity of ChoiceRule#validate down

kbrock commented:
update:
- added reject_field to reduce complexity from 8 => 5

kbrock commented:
update:
- reducing changed line count `+232/-110` => `+218/-78`

kbrock commented:
update:
- rebased onto newest name_path
- rearrange a few methods in validation mixin methods
- now passing workspace to all objects that need validation
- fixed 2 specs that had Next and End
-

kbrock commented:
WIP: This keeps starting small and growing large. Conflating too many ideas. Going to move the path code into the full name change.

This may become multiple PRs:

- unify error messages
- common helpers to make validation easier
- validate all things

kbrock commented:
I'm merging the dsl work into here and I squashed the commits down to the various forms of this code (did not keep the payload validator.)

- handle paths and complicated parsing
- handle every field (with larger datatype coverage)
- dsl to "simplify" some of the data concepts (having the object know the fields does makes validation and parsing easier)

This has run its course with a good chunk of the error message changes being promoted to #229
